pkg/signup/signup_handler.go

```diff
@@ -5,22 +5,31 @@ import (
 	"log"
 	"net/http"
 
-	"github.com/codeready-toolchain/registration-service/pkg/configuration"
 	"github.com/gin-gonic/gin"
+
+	"github.com/codeready-toolchain/registration-service/pkg/configuration"
 )
 
 // SignupService implements the signup endpoint, which is invoked for new user registrations.
 type SignupService struct {
 	config *configuration.Registry
 	logger *log.Logger
+	keyManager *KeyManager
 }
 
 // NewSignupService returns a new SignupService instance.
-func NewSignupService(logger *log.Logger, config *configuration.Registry) *SignupService {
-	return &SignupService{
+func NewSignupService(logger *log.Logger, config *configuration.Registry) (*SignupService, error) {
+	service := &SignupService{
 		logger: logger,
 		config: config,
 	}
+	// create new KeyManager
+	keyManager, err := NewKeyManager(logger, config)
```

Comment:
Instead of creating a new manager (which means we will fetch keys every time we create a service) for individual services we should use the only manager for all the services. We can initialize it in main and then re-use in all the services.
We would want to use the only instance of the manager for multiple reasons. Performance (don't fetch keys multiple times when not needed). Code-reusage. Also, our manager will need to evolve in the future. We will need to have some internal cache for the keys which we will update from time to time (keys could rotate in RHD). And having just one instance will be preferable for that.

Or we could keep introduce a singleton manager similar to what we use in fabric8-* services - https://github.com/fabric8-services/fabric8-common/blob/2cc2344d55776abe18178fc1c2efa8abc9996565/auth/token.go#L58
Then you don't have to pass the manager around. You just initialize it in main and then access everywhere via `auth.DefaultManager(config)`.

Reply:
I implemented this in f32824a, but I am not sure if that retrieval makes sense. The `func DefaultManager(config ManagerConfiguration)` uses `sync.Once`. That means repeated calls to it can be done, but have no effect. If that would be used to also *retrieve* the instance, then the caller always has to give a `config`, even if that is not used and has no effect on updated config values. I guess the rest of the code should retrieve the instance by directly accessing the `Manager` field, but still, any updated config and call to `DefaultManager()` would result in a silent ignore. That looks like a bogus design to me.

I updated this in 9560d40 to a different impl, that has an additional method to retrieve the KeyManager from global (and that one reminded me that golang has no polymophic method signatures, doh).